### PR TITLE
CORDA-1496 Add api-scanner documentation

### PR DESCRIPTION
**Pull request by anthonykeenan**
Add api scanner documentation to the docsite

### CODE REVIEW

File: docs/source/api-scanner.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+===================
+
+We have committed not to alter Corda's API so that developers will not have to keep rewriting their CorDapps with each
+new Corda release. The stable Corda modules are listed :doc:`here </corda-api>`. Our CI process runs an "API Stability"
+check for each GitHub pull request in order to check that we don't accidentally introduce an API-breaking change.
+
+Build Process
+-------------
+
+As part of the build process our the following commands are run for each PR:
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: Spurious word "our" here.

File: docs/source/api-scanner.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+.. code-block:: shell
+
+   $ gradlew generateApi
+   $ .ci/check-api-changes.sh
+
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: This `bash` script has been tested on both MacOS and various Linux distributions. (And it obviously doesn't work on Windows.)

File: docs/source/api-scanner.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Updating the API
+----------------
+
+As a rule, ``api-current.txt`` should only be updated by the release manager for each Corda release.
+
+We do not expect modifications to ``api-current.txt`` as part of normal development. However, we may sometimes need to adjust
+the public API in ways that would not break developers' CorDapps but which would be blocked by the API Stabilty check.
+For example, migrating a method from an interface into a superinterface. Any changes to the API summary file should be
+included in the PR, which would then need explicit approval from either Mike Hearn, Rick Parker or Matthew Nesbit.
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: If this document is going onto the Doc site then I don't think we need to mention Mike, Rick and Matthew :wink:.

File: docs/source/api-scanner.rst [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+As part of the build process our the following commands are run for each PR:
+
+.. code-block:: shell
+
+   $ gradlew generateApi
+   $ .ci/check-api-changes.sh
+
+The script's return value is the number of API-breaking changes that it has detected, and this should be zero for the
+check to pass. There are three kinds of breaking changes.
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: The maximum return value is 255. The script will still correctly _report_ higher numbers of breaking changes though.

File: docs/source/api-scanner.rst [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+----------------
+
+As a rule, ``api-current.txt`` should only be updated by the release manager for each Corda release.
+
+We do not expect modifications to ``api-current.txt`` as part of normal development. However, we may sometimes need to adjust
+the public API in ways that would not break developers' CorDapps but which would be blocked by the API Stabilty check.
+For example, migrating a method from an interface into a superinterface. Any changes to the API summary file should be
+included in the PR, which would then need explicit approval from either Mike Hearn, Rick Parker or Matthew Nesbit.
+
+.. note:: If you need to modify ``api-current.txt``, do not re-generate the file on the master branch. This will include new api that
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: Capitals for API, i.e. "... will include new API that ...".

File: docs/source/api-scanner.rst [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+   hasn't been released or committed to, and may be subject to change. Manually change the specific line or lines of the
+   existing committed api that has changed.
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: "... committed API ..."

File: docs/source/contributing.rst [hunk 7]
```diff
@@ -43,6 +43,12 @@ Building against the master branch
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 You can test your changes against CorDapps defined in other repos by following the instructions :doc:`here </building-against-master>`.
 
+Running the api scanner
+^^^^^^^^^^^^^^^^^^^^^^^
+Your changes must also not break compatibility with existing public api. We have an api scanner tool which runs as part of the build
```
Comment (chrisr3) [inline docs/source/contributing.rst]: "... existing public API..."

File: docs/source/api-scanner.rst (hunk 2 above)
Comment (anthonykeenan) [reply docs/source/api-scanner.rst]: It works for me in git bash.

Comment (chrisr3) [reply docs/source/api-scanner.rst]: OK, that's interesting. Maybe document that too?

**Comment by anthonykeenan** [comment]
And yes thanks for writing most of it for me ;-)

File: docs/source/contributing.rst [hunk 8]
```diff
@@ -45,7 +45,7 @@ You can test your changes against CorDapps defined in other repos by following t
 
 Running the api scanner
 ^^^^^^^^^^^^^^^^^^^^^^^
-Your changes must also not break compatibility with existing public api. We have an api scanner tool which runs as part of the build
+Your changes must also not break compatibility with existing public API. We have an api scanner tool which runs as part of the build
```
Comment (chrisr3) [inline docs/source/contributing.rst]: Is this "an API scanner tool" or "an `api-scanner` tool"?

File: docs/source/api-scanner.rst [hunk 9]
```diff
@@ -8,15 +8,19 @@ check for each GitHub pull request in order to check that we don't accidentally
...
 .. code-block:: shell
 
    $ gradlew generateApi
    $ .ci/check-api-changes.sh
 
-The script's return value is the number of API-breaking changes that it has detected, and this should be zero for the
-check to pass. There are three kinds of breaking changes.
+This `bash` script has been tested on both MacOS and various Linux distributions, it can also be run on Windows with the
+use of a suitable bash emulator such as git bash. The script's return value is the number of API-breaking changes that it
+has detected, and this should be zero for the check to pass. The maximum return value is 255, however, the script will still
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: ", however, " -> ", although"

File: docs/source/api-scanner.rst [hunk 10]
```diff
@@ -8,15 +8,19 @@ check for each GitHub pull request in order to check that we don't accidentally
...
+correctly report higher numbers of breaking changes.
+
+There are three kinds of breaking changes:
```
Comment (chrisr3) [inline docs/source/api-scanner.rst]: I think "kinds of breaking change" is better here.

File: docs/source/api-scanner.rst [hunk 11]
```diff
@@ -0,0 +1,58 @@
+Api Stability Check
+===================
+
+We have committed not to alter Corda's API so that developers will not have to keep rewriting their CorDapps with each
+new Corda release. The stable Corda modules are listed :doc:`here </corda-api>`. Our CI process runs an "API Stability"
```
Comment (joeldudleyr3) [inline docs/source/api-scanner.rst]: Should link to corda-api#internal-apis-and-stability-guarantees instead.

File: docs/source/api-scanner.rst [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+-------------
+
+As part of the build process the following commands are run for each PR:
+
+.. code-block:: shell
+
+   $ gradlew generateApi
+   $ .ci/check-api-changes.sh
+
+This `bash` script has been tested on both MacOS and various Linux distributions, it can also be run on Windows with the
```
Comment (joeldudleyr3) [inline docs/source/api-scanner.rst]: Did you mean to use double back-quotes instead?

File: docs/source/api-scanner.rst [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+correctly report higher numbers of breaking changes.
+
+There are three kinds of breaking changes:
+
+* Removal or modification of existing API, i.e. an existing class, method or field has been either deleted or renamed, or
+  its signature somehow altered.
+* Addition of a new method to an interface or abstract class. Types that have been annotated as ``@DoNotImplement`` are
+  excluded from this check. (This annotation is also inherited across subclasses and subinterfaces.)
+* Exposure of an internal type via a public API. Internal types are considered to be anything in a ``*.internal.`` package
+  or anything in a module that isn't in the stable modules list :doc:`here </corda-api>`.
```
Comment (joeldudleyr3) [inline docs/source/api-scanner.rst]: As above, link directly to corda-api#internal-apis-and-stability-guarantees

File: docs/source/api-scanner.rst [hunk 14]
```diff
@@ -0,0 +1,58 @@
+Api Stability Check
```
Comment (joeldudleyr3) [inline docs/source/api-scanner.rst]: General approach across docsite is not to capitalise each word in titles.

File: docs/source/contributing.rst [hunk 15]
```diff
@@ -43,6 +43,12 @@ Building against the master branch
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 You can test your changes against CorDapps defined in other repos by following the instructions :doc:`here </building-against-master>`.
 
+Running the api scanner
```
Comment (chrisr3) [inline docs/source/contributing.rst]: Same comment: "the API scanner" or "the `api-scanner`"?

**Pull request merged**